signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,804 @@
+            WstsNetMessageType::NonceResponse => self.signer_count(),
+            WstsNetMessageType::SignatureShareRequest => 1,
+            WstsNetMessageType::SignatureShareResponse => self.signer_count(),
+            _ => 0,
+        }
+    }
+
+    /// Gets the number of buffered messages of the given type that are
+    /// currently stored in this [`StateMachine`].
+    fn total_message_count(&self, message_type: WstsNetMessageType) -> u32 {
+        // The `_ as u32` should be safe here since we know that the number of
+        // signers is far less than `u32::MAX`, and each message is deduplicated
+        // by the sender's public key, which is also validated to be a valid
+        // member of the signer set.
+        self.wsts_messages
+            .keys()
+            .filter(|(msg_type, _)| *msg_type == message_type)
+            .count() as u32
+    }
+
+    /// Gets the number of signers that are expected to participate in the DKG
+    /// verification.
+    fn signer_count(&self) -> u32 {
+        self.coordinator.get_config().num_signers
+    }
+
+    /// Asserts that the [`StateMachine`] is in a state where it
+    /// can process messages.
+    fn assert_valid_state(&self) -> Result<(), Error> {
+        match self.state {
+            State::Expired => Err(Error::Expired),
+            State::Success(_) | State::Error => Err(Error::EndState(self.state.clone().into())),
+            State::Idle | State::Signing => Ok(()),
+        }
+    }
+
+    /// Determines the [`WstsNetMessageType`] of messages that can be processed
+    /// given the current state of the [`FrostCoordinator`].
+    ///
+    /// The allowed message types per state are as follows:
+    /// 1. If the coordinator is in the `Idle` state, it can only process
+    ///    `NonceRequest` messages.
+    /// 2. Upon receiving a `NonceRequest`, the coordinator transitions to the
+    ///    `NonceGather` state. In this state, the coordinator can process
+    ///    `NonceResponse` messages until it has received a number of nonces
+    ///    equal to the number of signers.
+    /// 3. Once this condition is met, the coordinator transitions to the
+    ///    `SignatureShareRequest` state, after which it can process
+    ///    `SignatureShareResponse` messages until it has received a number of
+    ///    signature shares equal to the number of signers.
+    /// 4. Once this condition is met, the coordinator will transition to either
+    ///    the `Success` or `Error` state, depending on the result of the
+    ///    signing operation.
+    fn current_processable_message_type(&self) -> WstsNetMessageType {
+        match self.coordinator.state {
+            wsts::state_machine::coordinator::State::Idle => WstsNetMessageType::NonceRequest,
+            wsts::state_machine::coordinator::State::NonceGather(_) => {
+                WstsNetMessageType::NonceResponse
+            }
+            _ => WstsNetMessageType::SignatureShareResponse,
```

Comment:
As the underlying coordinator can end up also in other states (eg, dkg related ones), can we specify the exact state for `SignatureShareResponse` and return an err for the others?

Reply:
This was added in the bigger [c371a97](https://github.com/stacks-network/sbtc/pull/1324/commits/c371a974ef39859a568921efd032ea9ac274abb3)